### ISSUE TITLE
Safe persistence and migration of gateway federation config

### ISSUE THREAD

**Issue by okjodom** [fedimint/fedimint#1057]
Assuming Federation ID's are stable, we should find a safe way of persisting gateway configs, without a direct and immediate overwrite of any previously saved configs that belong to the same Federation ID

**Comment by elsirion** [comment]
Do we actually ever need to overwrite config? Shouldn't joining the same federation a) fail or b) try to start the actor if not already running? 

A more important, but different problem is that we currently use the federation name, something that is easily forgeable. If it was the hash over a public key set we could require a signature over the config by that key set. Only if it's valid we'd accept the config. Once we allow config to change we might want to allow overwriting in that case.

**Comment by okjodom** [comment]
> Do we actually ever need to overwrite config?

Yes. But I would call it **migrate config**. AFAIK, the saved config is a source of truth representation of a federation, proxied by the actor. A migrate config scenario  would mean the federation has changed, and we are notifying a previously linked gateway of this update to the federation. This assumes:
- The new federation maintains the same stable identifier: #754 
- The notification is a simple reuse of the `register(connect-info)` API

> Shouldn't joining the same federation a) fail or b) try to start the actor if not already running?

Maybe we should:
- Deep compare the saved config against proposed config, asserting that we actually have a change.
- In case of some change, (_and we can't somehow upgrade and rehydrate an existing actor?_), we should care to:
  - stop accepting new transactions to now old federation actor
  - clear all in flight tx
  - peg out for the gateway? (_but we should have done this earlier!_)

1. Migration experience should ensure no fund loss for the gateway
2. De-registering a federation should be an explicit action by the gateway admin.

**Comment by okjodom** [comment]
> A more important, but different problem is that we currently use the federation name, something that is easily forgeable. If it was the hash over a public key set we could require a signature over the config by that key set. Only if it's valid we'd accept the config. Once we allow config to change we might want to allow overwriting in that case.

Deferring to #754

**Comment by m1sterc001guy** [comment]
@okjodom @elsirion do you guys think this is still an issue? After the migration to the new client, we no longer have actors and we just keep a map of Federation ID -> Client. Each client's config is stored in a database and read when `gatewayd` first starts.

If `connect_federation` is called again for the same client, the gateway does currently go off and download the client config from the federation and re-write it to the database. Do you guys think we should add a check here? I was thinking that its fine to overwrite the existing config, a gateway admin might want to do this if they're getting connection errors or something.

**Comment by justinmoon** [comment]
> If connect_federation is called again for the same client, the gateway does currently go off and download the client config from the federation and re-write it to the database.

I don't think this is a problem.

I don't think this is relevant anymore.
